### PR TITLE
[Test] unittests for Utility.class.inc

### PR DESCRIPTION
**Pull request by AlexandraLivadas**
### Brief summary of changes
I have started writing unit tests for the Utility class on this branch. 

### This is related to:
#4378 

### To test this change...
Run "npm run tests:unit -- --filter Utility" or check out Travis

### CODE REVIEW

**Comment by PapillonMcGill** [comment]
Hello Alexandra,
I haven't tested it yet, it does look good.
`sudo` should not be needed to run this test. It is also a security risk to run this test with sudo.

File: php/libraries/Utility.class.inc [hunk 1]
```diff
@@ -28,7 +28,7 @@ class Utility
     /**
      * Stops anyone from instantiating an instance of Utility
      */
-    private function __construct()
+    public function __construct()
```
Comment (driusan) [inline php/libraries/Utility.class.inc]: There's no reason to make the Utility class constructable. It's holds a bunch of static functions, there is no instance data to go with it.

Comment (AlexandraLivadas) [reply php/libraries/Utility.class.inc]: Hi! Sorry about that I misunderstood how the Utility class was organized and how the methods were to be called. Just pushed my edits!

**Comment by kongtiaowang** [comment]
It failed on Travis, because of VisitTest::testVisitsProjects. I will fix this.

**Comment by AlexandraLivadas** [comment]
Hi! I just pushed some changes. If you run the tests you'll notice that two of them are marked incomplete. That is because a lot of the methods in the Utility class use `\Database::singleton()` rather than `NDB_Factory::singleton()` and I am having a hard time figuring out how to correctly mock and test that.

I've tried using similar logic to the rest of my tests but the error it gives me is "DatabaseException: No database connection exists"! I'll continue to look around for answers, but does anyone happen to know more about this?

**Comment by johnsaigle** [comment]
@AlexandraLivadas I've definitely come across it before but I don't really know how to fix it :( Sorry I can't be of more help.

The rest of the tests look good!

**Comment by kongtiaowang** [comment]
@AlexandraLivadas 
This https://github.com/aces/Loris/pull/438 may help you to figure out how to setup mock DB.

File: test/unittests/UtilityTest.php [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+ * Unit test for Utility class
+ * @category Tests
+ * @package  Main
+ * @author   Alexandra Livadas <alexandra.livadas@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+class FakeUtility extends Utility 
+{
```
Comment (kongtiaowang) [inline test/unittests/UtilityTest.php]: Missing comment makes phan checking failed.
    /**
     * Returns a list of sites in the database
     *
     * @param bool $study_site If true only return sites that are
     *                         study sites according to the psc
     *                         table
     *
     * @return array an associative array("center ID" => "site name")
     */

File: test/unittests/UtilityTest.php [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+        $result = $DB->pselect($query, array());
+        // fix the array
+        $list = array();
+        foreach ($result as $row) {
+            $list[$row["CenterID"]] = $row["Name"];
+        }
+        natcasesort($list);
+        return $list;
+    }
+
```
Comment (kongtiaowang) [inline test/unittests/UtilityTest.php]: same here.

File: test/unittests/UtilityTest.php [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+    {
+        $query = "SELECT Visit_label from Visit_Windows ORDER BY Visit_label";
+        $result = $DB->pselect($query, array());
+        $list = array();
+        foreach ($result as $row) {
+            $list[$row["Visit_label"]] = ucfirst($row["Visit_label"]);
+        }
+        return $list;
+    }
+
```
Comment (kongtiaowang) [inline test/unittests/UtilityTest.php]: same here

File: test/unittests/UtilityTest.php [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+            "SELECT Test_name,Full_name FROM test_names WHERE IsDirectEntry=true 
+             ORDER BY Full_name",
+            array()
+        );
+        foreach ($instruments_q as $key) {
+            $instruments[$key['Test_name']] =$key['Full_name'];
+        }
+        return $instruments;
+    }
+
```
Comment (kongtiaowang) [inline test/unittests/UtilityTest.php]: same

**Comment by kongtiaowang** [comment]
test/unittests/UtilityTest.php:37 PhanParamSignatureMismatch Declaration of function getSiteList(\Database $DB, bool $study_site = null) : array should be compatible with function getSiteList(bool $study_site = null) : array defined in php/libraries/Utility.class.inc:127
test/unittests/UtilityTest.php:37 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getSiteList(\Database $DB, bool $study_site = null) : array should be compatible with function getSiteList(bool $study_site = null) : array (the method override requires 1 parameter(s), but the overridden method requires only 0) defined in php/libraries/Utility.class.inc:127
test/unittests/UtilityTest.php:60 PhanParamSignatureMismatch Declaration of function getVisitList(\Database $DB) : array|array<string,string> should be compatible with function getVisitList() : array<string,string> defined in php/libraries/Utility.class.inc:191
test/unittests/UtilityTest.php:60 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getVisitList(\Database $DB) : array should be compatible with function getVisitList() : array (the method override requires 1 parameter(s), but the overridden method requires only 0) defined in php/libraries/Utility.class.inc:191
test/unittests/UtilityTest.php:78 PhanParamSignatureMismatch Declaration of function getDirectInstruments(\Database $DB) : array should be compatible with function getDirectInstruments() : array defined in php/libraries/Utility.class.inc:576
test/unittests/UtilityTest.php:78 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getDirectInstruments(\Database $DB) : array should be compatible with function getDirectInstruments() : array (the method override requires 1 parameter(s), but the overridden method requires only 0) defined in php/libraries/Utility.class.inc:576
test/unittests/UtilityTest.php:100 PhanParamSignatureMismatch Declaration of function getStageUsingCandID(\Database $db, string $Cand_id) : string should be compatible with function getStageUsingCandID(string $Cand_id) : string defined in php/libraries/Utility.class.inc:627
test/unittests/UtilityTest.php:100 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getStageUsingCandID(\Database $db, string $Cand_id) : string should be compatible with function getStageUsingCandID(string $Cand_id) : string (the method override requires 2 parameter(s), but the overridden method requires only 1) defined in php/libraries/Utility.class.inc:627
test/unittests/UtilityTest.php:116 PhanParamSignatureMismatch Declaration of function getSubprojectIDUsingCandID(\Database $db, string $Cand_id) : int should be compatible with function getSubprojectIDUsingCandID(string $Cand_id) : int defined in php/libraries/Utility.class.inc:644
test/unittests/UtilityTest.php:116 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getSubprojectIDUsingCandID(\Database $db, string $Cand_id) : int should be compatible with function getSubprojectIDUsingCandID(string $Cand_id) : int (the method override requires 2 parameter(s), but the overridden method requires only 1) defined in php/libraries/Utility.class.inc:644
test/unittests/UtilityTest.php:130 PhanParamSignatureMismatch Declaration of function getTestNameUsingFullName(\Database $db, string $fullname) : string should be compatible with function getTestNameUsingFullName(string $fullname) : string defined in php/libraries/Utility.class.inc:663
test/unittests/UtilityTest.php:130 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getTestNameUsingFullName(\Database $db, string $fullname) : string should be compatible with function getTestNameUsingFullName(string $fullname) : string (the method override requires 2 parameter(s), but the overridden method requires only 1) defined in php/libraries/Utility.class.inc:663
test/unittests/UtilityTest.php:149 PhanParamSignatureMismatch Declaration of function getLanguageList(\Database $DB) : array should be compatible with function getLanguageList() : array defined in php/libraries/Utility.class.inc:905
test/unittests/UtilityTest.php:149 PhanParamSignatureRealMismatchTooManyRequiredParameters Declaration of function getLanguageList(\Database $DB) : array should be compatible with function getLanguageList() : array (the method override requires 1 parameter(s), but the overridden method requires only 0) defined in php/libraries/Utility.class.inc:905
test/unittests/UtilityTest.php:448 PhanParamTooMany Call with 2 arg(s) to \FakeUtility::getTestNameByCommentID() which only takes 1 arg(s) defined at php/libraries/Utility.class.inc:282
test/unittests/UtilityTest.php:449 PhanParamTooMany Call with 2 arg(s) to \FakeUtility::getTestNameByCommentID() which only takes 1 arg(s) defined at php/libraries/Utility.class.inc:282
[debug] Restarted process exited 1

**Comment by johnsaigle** [comment]
@AlexandraLivadas Shen posted above some errors generated by `phan`, our static analysis tool. The tool indicates some potential problems in the code and so we won't merge your change until they're remedied.

To replicate them locally, you can run `make checkstatic`.

Let me know if you have any questions about how to address them.

**Comment by kongtiaowang** [comment]
Try this to test GetSubprojectListWithProjectID method.
public function testGetSubprojectListWithProjectID()
   {
       $this->_dbMock->expects($this->any())
           ->method('pselect')
           ->willReturn(
               array (
                   array('SubprojectID' => '123',
                         'title'        => 'DemoProject',

                   )
               ));
       $this->assertEquals(
           array('123' => 'DemoProject'),
           FakeUtility::getSubprojectList('123'));
   }

**Comment by kongtiaowang** [comment]
Fixing the phan error, in this case, make a fake method name to test it.

File: test/unittests/UtilityTest.php [hunk 6]
```diff
@@ -0,0 +1,1042 @@
...
+        $query = "SELECT Visit_label from Visit_Windows ORDER BY Visit_label";
+        $result = $DB->pselect($query, array());
+        $list = array();
+        foreach ($result as $row) {
+            $list[$row["Visit_label"]] = ucfirst($row["Visit_label"]);
+        }
+        return $list;
+    }
+    /**
+     * Gets a list of all instruments where are administered as direct data
```
Comment (PapillonMcGill) [inline test/unittests/UtilityTest.php]: Could you rephrase this line? It is hard to decipher.
Also, if using end ending dots, should use for all the phrase.

**Comment by PapillonMcGill** [comment]
@AlexandraLivadas In the summary of this PR, could you remove "sudo" from the "**To test this change...**". It is not needed and pose a security risk,

File: test/unittests/UtilityTest.php [hunk 7]
```diff
@@ -368,6 +368,34 @@ static function fakeGetTestNameByCommentID(
...
     }
 
+    /**
+     * Get a list of DDE instruments installed in Loris.
+     *
+     * @param \NDB_Config $config The mock config used for testing
+     *
+     * @return array of the form Test_name => Full Description
+     */
+    static function fakeGetAllDDEInstruments(\NDB_Config $config): array
```
Comment (PapillonMcGill) [inline test/unittests/UtilityTest.php]: I thought that FakeUtility was scrap.  Testing a copy of a class do not test the class itself but a copy of the class at some point in time that is not necessarily in sync.

Comment (AlexandraLivadas) [reply test/unittests/UtilityTest.php]: Based on my presentation today and on the new issue #4989 I'm going to try to scrap it today or tomorrow! Stay tuned :)

**Comment by AlexandraLivadas** [comment]
For my latest commit: I've updated a couple of the Utility methods that use `$DB =& \Database::singleton();` and their related unit tests. I haven't yet deleted the "fake" versions of the methods from the FakeUtility class just in case!

**Comment by christinerogers** [comment]
> This is related to: Issue #4989
Nice! 

> I haven't yet deleted the "fake" versions of the methods from the FakeUtility class just in case!

Ok - Could you add comments in the code to identify these for later cleanup, if they are to stay for now?  (whatever the appropriate version of "fixme" is)

File: test/unittests/UtilityTest.php [hunk 8]
```diff
@@ -0,0 +1,1230 @@
...
+        $this->_dbMock->expects($this->any())
+            ->method('pselect')
+            ->willReturn($test_names);
+        $this->_configMock->expects($this->any())
+            ->method('getSetting')
+            ->willReturn($doubleDataEntryInstruments);
+        //$this->markTestIncomplete("This test is incomplete!");
+        $this->assertEquals(
+            array('test_name2' => 'full_name2'),
+            FakeUtility::fakeGetAllDDEInstruments($this->_configMock)
```
Comment (PapillonMcGill) [inline test/unittests/UtilityTest.php]: Still call FakeUtility.

and others place below.

**Comment by PapillonMcGill** [comment]
Is it ready for review? Can we remove the 'need work' label?

**Comment by AlexandraLivadas** [comment]
> Is it ready for review? Can we remove the 'need work' label?

The  'need work' label was added by @johnsaigle, I think because it was failing on Travis. I think it can be removed @johnsaigle @kongtiaowang ?
